### PR TITLE
Prepare publication of the nb-javac to maven central

### PR DESCRIPTION
**Pull request by matthiasblaesing**
- modified build script to include publishing to sonatype OSSRH snapshot
  repository and to the OSSRH staging repository for maven central
- added documentation about the publication process

Signed-of-by: Matthias Bläsing <mblaesing@doppel-helix.eu>

### CODE REVIEW

**Comment by matthiasblaesing** [comment]
This is a followup/part of a fix for [NETBEANS-5109](https://issues.apache.org/jira/browse/NETBEANS-5109). I implemented maven central publication for JNA a long time ago and reused parts of that procedure to make it possible to publish nb-javac to maven central.

This PR alone is not enough - the nb-javac team will still need to adjust the pom to create the right person information and update the groupId in the template pom, get accounts from sonatype and do the publishing, but it is the part I can contribute.

I updated the documentation to contain description about the process and I'm willing to give some more info about the process where required.

I used my personal group id to create a snapshot (no I won't publish javac) to demonstrate the result of this PR:

https://oss.sonatype.org/content/repositories/snapshots/eu/doppel-helix/com/github/oracle/nb-javac/nb-javac/15-SNAPSHOT/

As you can see, the "-impl.jar" seems to be missing, but that is not the case, maven requires a base artifact and so the implementation became https://oss.sonatype.org/content/repositories/snapshots/eu/doppel-helix/com/github/oracle/nb-javac/nb-javac/15-SNAPSHOT/nb-javac-15-20201218.220825-1.jar. 

The "-api.jar" is bundled with a different classifier: https://oss.sonatype.org/content/repositories/snapshots/eu/doppel-helix/com/github/oracle/nb-javac/nb-javac/15-SNAPSHOT/nb-javac-15-20201218.220825-1-api.jar

The javadoc, that is present on maven central is only a place holder, to satisfy the requirements of sonatype.

I thought about creating two artifacts, but I think this is the cleaner solution as this is more or less a direct transfer from the current publication details.

File: make/langtools/netbeans/nb-javac/build.xml
Comment (JaroslavTulach) [inline make/langtools/netbeans/nb-javac/build.xml:143]: typo: `ot`.

File: make/langtools/netbeans/nb-javac/pom-nb-javac.xml
Comment (JaroslavTulach) [inline make/langtools/netbeans/nb-javac/pom-nb-javac.xml:7]: Could this line be replaced with a `TEMPLATE` and processed just like the `<version/>` tag in the build script? The publishing script would then have to be invoked with `ant -DgroupId=eu.doppel-helix.com.github.oracle.nb-javac` (or some other group ID). As a benefit, the whole publishing process would be exactly recorded in the repository and anyone with proper OOSRH account could replay it.

File: make/langtools/netbeans/nb-javac/pom-nb-javac.xml
Comment (JaroslavTulach) [inline make/langtools/netbeans/nb-javac/pom-nb-javac.xml:37]: Congratulation on becoming `nb-javac` developer!

I assume your entry should stay as you have contributed the maven deployment script. In addition to that contacts to @Akshay-Gupta-Oracle, @singh-akhilesh or some generic contact to maintainers of this repository shall be added as well.

File: make/langtools/netbeans/nb-javac/build.xml
Comment (JaroslavTulach) [inline make/langtools/netbeans/nb-javac/build.xml:149]: If the `groupId` is turned into a template, this line would become:
```
ant publish-to-ossrh-snapshots -DgroupId=your.group.id
```
and the line below would be
```
ant publish-to-maven-central -DgroupId=your.group.id
```

**Review by JaroslavTulach — CHANGES_REQUESTED** [review]
Thanks a lot Matthias for resolving the technical problems of uploading `nb-javac` properly to Maven central. Once your PR is merged, it would be just a matter of getting the account and handling the upload. I'd prefer if the `groupId` wan't hardcoded in the sources to make it clear the upload is open to everyone.

File: make/langtools/netbeans/nb-javac/pom-nb-javac.xml
Comment (matthiasblaesing) [reply make/langtools/netbeans/nb-javac/pom-nb-javac.xml:37]: No I should not. I won't publish nb-javac, but being listed in the POM is a requirement to be able to publish to maven central. I tested publishing to the snapshot repository and the staging repository and for that to work, I had to be listed. Just as the groupId is currently not usable for the oracle team, this needs to be updated accordingly. 

That is also the reason, that I don't intent to make the groupId a variable.  The pom-nb-javac.xml is a configuration file per group that redistributes. The only thing that regularly changes over time is the version, the developers will be pretty constant and the same is true for the group id.

File: make/langtools/netbeans/nb-javac/build.xml
Comment (matthiasblaesing) [reply make/langtools/netbeans/nb-javac/build.xml]: Fixed in place.

File: make/langtools/netbeans/nb-javac/pom-nb-javac.xml
Comment (matthiasblaesing) [reply make/langtools/netbeans/nb-javac/pom-nb-javac.xml:7]: Please see my comment on the developer section.

File: make/langtools/netbeans/nb-javac/pom-nb-javac.xml
Comment (JaroslavTulach) [reply make/langtools/netbeans/nb-javac/pom-nb-javac.xml:37]: According to the first paragraph the `nb-javac` maintainers are supposed to __replace__ the developer section with their own information. OK, let them do it.

File: make/langtools/netbeans/nb-javac/pom-nb-javac.xml
Comment (JaroslavTulach) [reply make/langtools/netbeans/nb-javac/pom-nb-javac.xml:7]: > ... I don't intent to make the groupId a variable. ... the developers will be pretty constant 
> and the same is true for the group id.

I see. Then your PR needs to be modified before it gets to `origin/master`. Can I push my preferred changes into this PR or shall I create my own and use bce95d8 as base?

Comment (matthiasblaesing) [reply make/langtools/netbeans/nb-javac/pom-nb-javac.xml:7]: Sure - this was intended to get the process started and demonstrate, that it is possible with too much fuzz. Feel free to improve, it's appreciated.

**Comment by JaroslavTulach** [comment]
#1 got merged and now there are conflicts in the `README.md`

File: make/langtools/netbeans/nb-javac/pom-nb-javac.xml
Comment (JaroslavTulach) [reply make/langtools/netbeans/nb-javac/pom-nb-javac.xml:7]: There is [PR #1](https://github.com/matthiasblaesing/nb-javac/pull/1) with my changes to make `groupId` pluggable.

**Comment by matthiasblaesing** [comment]
I rebased my change onto current master. To get this further, action from the oracle team is needed:

- a groupId needs to be acquired (I don't know the agreements between Oracle and Sonatype or Oracle internal, so I can't really help there) (=> https://github.com/oracle/nb-javac/pull/2/files#diff-4567d841ae351146e993ae3fa61d189ef40a65e7d69da8d31c83720b9f5163e0R7)
- the first set of developers that will do the publishing need to be determined (=> https://github.com/oracle/nb-javac/pull/2/files#diff-4567d841ae351146e993ae3fa61d189ef40a65e7d69da8d31c83720b9f5163e0R30-R39)

With the above to infos we (I can do that) need to update the `pom-nb-javac.xml` to its final contents. After the pom is updated and this PR is merged, a first maven central publication is possible.

**Comment by matthiasblaesing** [comment]
@JaroslavTulach Thank you for the update - I merged your changes into this PR. I also updated the pom template to hold the current main contributors and fixed an issue when the source zip was created independently.

File: make/langtools/netbeans/nb-javac/pom-nb-javac.xml
Comment (JaroslavTulach) [inline make/langtools/netbeans/nb-javac/pom-nb-javac.xml:41]: The `h` at the end is probably a typo.

**Review by JaroslavTulach — APPROVED** [review]
Amazing. This PR seems to have converged to a solution that fully reflects existing reality and allows independent parties to perform the upload. I support merging and seeking for such party to release `nb-javac-15.1`.

**Comment by matthiasblaesing** [comment]
Thank you for the review and contributions @JaroslavTulach. I fixed the typo and I think is ready to merge @Akshay-Gupta-Oracle

**Comment by JaroslavTulach** [comment]
So am I. @lahodaj, would you agree?

**Review by lahodaj — APPROVED** [review]
Looks reasonable to me.

**Review by singh-akhilesh — APPROVED** [review]
:thumbsup:

**Comment by arvindaprameya** [comment]
Approved, Thanks for this contribution Matthias .

**Comment by matthiasblaesing** [comment]
Thank you all for review. This needs a commiter (I'm not) and then someone to get a group id and get this published. Once that has happend I'm willing to prepare a PR against netbeans to move to the maven central artifact.

**Review by Akshay-Gupta-19 — APPROVED** [review]
Looks correct.

File: make/langtools/netbeans/nb-javac/pom-nb-javac.xml
Comment (JaroslavTulach) [inline make/langtools/netbeans/nb-javac/pom-nb-javac.xml:1]: @matthiasblaesing , this file doesn't have proper license header file and gets included on Maven central. Don't you want to create a PR to add the license?

**Review by aidevbd24 — APPROVED** [review]

**Pull request merged**
